### ISSUE TITLE
error_reporting not passed to PHP process when Psalm is restarted without Xdebug

### ISSUE THREAD

**Issue by sebastianbergmann** [vimeo/psalm#7860]
When I use Psalm from its PHAR and Xdebug is loaded then I get these deprecation warnings:
```
PHP Deprecated:  Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 15
Deprecated: Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 15
PHP Deprecated:  Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 19
Deprecated: Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 19
```

I do not see these deprecation warnings when Xdebug is disabled.

Based on the dependency on `composer/xdebug-handler` that is listed in Psalm's `composer.json`, I assume that Psalm restarts in a PHP process without Xdebug when Xdebug is detected. It appears that this PHP process does not use the same `error_reporting` configuration as the original PHP process.

#### PHP Version
```
$ php -v                                                           
PHP 8.1.4 (cli) (built: Mar 15 2022 20:18:41) (NTS gcc x86_64)
Copyright (c) The PHP Group
Zend Engine v4.1.4, Copyright (c) Zend Technologies
```

#### PHP Error Reporting
```
$ php -i | grep error_reporting
error_reporting => 24575 => 24575
```
`24575` is equivalent to `E_ALL & ~E_DEPRECATED`.

#### Psalm Version
```
$ ./tools/psalm --version                 
Psalm 4.x-dev@
```
`4.x-dev@` is not a correct version identifier, but that is the topic of another [issue](https://github.com/vimeo/psalm/issues/7606).
```
$ grep psalm .phive/phars.xml 
<phar name="psalm" version="^4.0" installed="4.22.0" location="./tools/psalm" copy="true"/>
```

#### Without Xdebug
```
$ ./tools/psalm --config=.psalm/config.xml
Target PHP version: 8.1 (inferred from composer.json)
Scanning files...
Analyzing files...

------------------------------
       No errors found!       
------------------------------
14 other issues found.
You can display them with --show-info=true
------------------------------

Checks took 0.01 seconds and used 18.699MB of memory
No files analyzed
Psalm was able to infer types for 99.0310% of the codebase
```

#### With Xdebug
```
$ ./tools/psalm --config=.psalm/config.xml
PHP Deprecated:  Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 15
Deprecated: Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 15
PHP Deprecated:  Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 19
Deprecated: Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 19
Target PHP version: 8.1 (inferred from composer.json)
Scanning files...
Analyzing files...

------------------------------
       No errors found!       
------------------------------
14 other issues found.
You can display them with --show-info=true
------------------------------

Checks took 0.01 seconds and used 18.712MB of memory
No files analyzed
Psalm was able to infer types for 99.0310% of the codebase
```
The deprecations reported are not in Psalm's own code, but in [humbug/box](https://github.com/vimeo/psalm/issues/7314).

**Comment by orklah** [comment]
Psalm is pretty strict with error_reporting. The first thing done after disabling GC is to set the error_reporting to `E_ALL | E_STRICT` (which is quite redundant since PHP 5.5 btw...)
https://github.com/vimeo/psalm/blob/b924032850df54dd05ffae409a709820fd6e688e/src/Psalm/Internal/ErrorHandler.php#L56

But that should happen whether or not Psalm is restarted. As far as I can tell, the only moment when Psalm loosen its error_reporting level is when it's using Reflection to scan classes :
https://github.com/vimeo/psalm/blob/35577dfd28e684391eb3021fb6d681cfff01de5b/src/Psalm/Internal/Codebase/Scanner.php#L704

Could you run with --debug to see if we can locate when it crashes?

**Comment by sebastianbergmann** [comment]
I see no difference in output when I use `--debug`:
```
$ ./tools/psalm --config=.psalm/config.xml --debug                
PHP Deprecated:  Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 15
Deprecated: Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::getIterator() should either be compatible with IteratorAggregate::getIterator(): Traversable, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 15
PHP Deprecated:  Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 19
Deprecated: Return type of HumbugBox383\KevinGH\RequirementChecker\RequirementCollection::count() should either be compatible with Countable::count(): int, or the #[\ReturnTypeWillChange] attribute should be used to temporarily suppress the notice in phar:///usr/local/src/type/tools/psalm/.box/src/RequirementCollection.php on line 19
Target PHP version: 8.1 (inferred from composer.json)
Scanning files...
0 changed files: 
    
Analyzing files...
Checking class references

------------------------------
       No errors found!       
------------------------------
16 other issues found.
You can display them with --show-info=true
------------------------------

Checks took 0.01 seconds and used 18.706MB of memory
No files analyzed
Psalm was able to infer types for 99.0440% of the codebase
```

**Comment by weirdan** [comment]
@sebastianbergmann can you reproduce it with a *non*-Phar release? Do you specify custom autoloader in Psalm config (those can change error reporting and install their own error handler)?

**Comment by sebastianbergmann** [comment]
This is the project where I am experiencing this: https://github.com/sebastianbergmann/type, https://github.com/sebastianbergmann/type/blob/main/.psalm/config.xml is the Psalm configuration used.

**Comment by weirdan** [comment]
So no autoloaders then. Can you try a non-phar release?

**Comment by sebastianbergmann** [comment]
> Can you try a non-phar release?

I could, but I think that would be pointless: the deprecations reported are not in Psalm's own code but in [humbug/box](https://github.com/vimeo/psalm/issues/7314), which is used to create Psalm's PHAR.

**Comment by weirdan** [comment]
Deprecation messages have been fixed in #7314
